docs/exceptions.md

```diff
@@ -0,0 +1,281 @@
+```
+
+### Pure code
+
+Pure code is one that does not use `IO` operations. For example:
+
+```
+lookupMaybe :: Map k v -> k -> Maybe v
+```
+
+Notice that while it is conceivable that we have a `Map` of `IO` actions, and
+instantiate `lookupMaybe` to `Map k (IO ()) -> k -> Maybe (IO ())`, the
+instantiation does not inspect/use these `IO` actions, so it's not considered
+potentially impure.
+
+### Identifying code purity
+
+Is there an instantiation of type parameters that would mean that there are
+`IO` actions inspected/used in the code?
+
+* Yes: it is potentially impure code
+* No: it is 100% pure code
+
+## Error handling practices
+
+### Pure code, programmer mistakes
+
+Do *not*:
+
+* do this often, as if it's okay
+* use `undefined`
+* use non-exhaustive pattern matching
+
+Do:
+
+* try to use types to avoid the need in the first place
+* comment extensively (invariants and precondition, reasoning)
+* use the `error` function (or `impureThrow` with custom exception)
+* use the `HasCallStack` feature
+
+Before writing code that allows programmer mistakes, consult with colleagues how
+it would be possible to prevent them statically. There might be tricks that
+you're not aware of. Always make this trade-off conciously and responsibly.
+
+(On the other hand, if static guarantees require GADTs or type families, perhaps
+it's better to not overcomplicate code. Use your judgement.)
+
+DISCUSSION: Should we create a synonym `bug = impureThrow` in Universum? This
+would make the intention more clear.
+
+### Pure code, regular errors
+
+Do *not*:
+
+* use `error` or `impureThrow`
+* use `MonadFail`
+* return `Either Text`
+
+Do:
+
+* return `Either ErrorADT`, `Maybe`
+* wrap the underlying (pure!) monad in `ExceptT` or `CatchT`
+* use `MonadError` or `MonadThrow` (methods of these classes). Note
+  that if you define `f :: MonadError m => m ()`, it won't be pure
+
+Consider parsing: it is pure, but we cannot make assumptions about the input. In
+this case we might want to use `ExceptT ParseError`. Or consider a lookup in a
+`Map`, where we don't know whether the key is present -- in this case we'd like
+to return `Maybe v`. In 100% pure code, use one of these ways to handle errors:
+
+* `Maybe`, `Either e`
+* `MaybeT`, `ExceptT e`
+* `CatchT`
+
+Avoid using `Text` with the error message in place of `e` -- create a proper
+ADT. In case creating a proper ADT feels too cumbersome, use `CatchT`, which
+is equivalent to `ExceptT SomeException`.
+
+Be careful not to use `MaybeT`, `ExceptT`, and `CatchT` in potentially impure
+code. When in doubt whether the code is potentially impure, use `MonadThrow`.
+(The reason we don't want `ExceptT` and co. in potentially impure code is that
+they add additional exception mechanisms to the one that `IO` has, and
+`catch`/`bracket` don't account for this).
+
+### Impure code, regular errors
+
+Do *not*:
+
+* use `error` or `impureThrow`
+* use `ExceptT`, `MaybeT`, or `CatchT`
+* use `Maybe` or `Either`
```

Comment:
1. What's wrong with `Maybe`?
2. AFAIU, you mean that `Either` shouldn't be used to represent regular errors, but it's fine to return `Either SomethingThatIsNotAnError SomethingElseThatIsNotAnError `, is it correct?

Reply:
Example: is it ok to return `Block`? Currently `Block` is `Either GenesisBlock MainBlock`. Of course it can be changed to a new type which is not `Either`, but suppose we don't want to do it for some reason. For example, because it's not trivial and will take some time, but one may want to return `Block` from a function right now.

Reply:
The idea was to have uniform throwing principle for potentially impure code (i.e. always throw an exception with `throwM`). Now that we plan to relax this rule to allow `Either NonExc NonExc`, then allowing `Maybe` makes sense as well.